libraries/app/api.cpp

```diff
@@ -134,23 +134,65 @@ namespace graphene { namespace app {
 
     void network_broadcast_api::on_applied_block( const signed_block& b )
     {
-       if( _callbacks.size() )
-       {
-          /// we need to ensure the database_api is not deleted for the life of the async operation
-          auto capture_this = shared_from_this();
-          for( uint32_t trx_num = 0; trx_num < b.transactions.size(); ++trx_num )
+       /// we need to ensure the database_api is not deleted for the life of the async operation
+       auto capture_this = shared_from_this();
+
+       fc::async( [this,capture_this,b]() {
+          int32_t block_num = int32_t(b.block_num());
+          if( _callbacks.size() )
           {
-             const auto& trx = b.transactions[trx_num];
-             auto id = trx.id();
-             auto itr = _callbacks.find(id);
-             if( itr != _callbacks.end() )
+             for( int32_t trx_num = 0; trx_num < b.transactions.size(); ++trx_num )
              {
-                auto block_num = b.block_num();
-                auto& callback = _callbacks.find(id)->second;
-                fc::async( [capture_this,this,id,block_num,trx_num,trx,callback](){ callback( fc::variant(transaction_confirmation{ id, block_num, trx_num, trx}) ); } );
+                const auto& trx = b.transactions[trx_num];
+                auto id = trx.id();
+                auto itr = _callbacks.find(id);
+                if( itr == _callbacks.end() ) continue;
+                confirmation_callback callback = itr->second;
+                itr->second = [](variant){};
+                callback( fc::variant(transaction_confirmation{ id, block_num, trx_num, false }) );
+                _callbacks.erase( itr );
+                auto tx_itr = _pending_tx.find( id );
+                if( tx_itr != _pending_tx.end() )
+                  _pending_tx.erase( tx_itr );
              }
           }
-       }
+
+          /// clear all expirations
+          while( true )
+          {
+             auto exp_it = _callbacks_expirations.begin();
+             if( exp_it == _callbacks_expirations.end() )
+                break;
+             if( exp_it->first >= b.timestamp )
+                break;
```

Comment:
I believe with careful thought about iterators and ``erase()``, this ``while`` loop could be changed to a ``for``. That could eliminate these ``if`` statements.

Reply:
On second thought, I'm thinking that this iterator requires a lock for all threads, the whole time of the iteration, due to erasures on other threads. This would turn this process into a single threaded one. Therefore, I propose a final thread, that is passed the futures of all these asyncs, joins them, and when complete, handles the erasures. This provides for multithreaded performance gains with only 1 thread locking to erase all callbacks tied to expirations.

These threads still require "read locks" due to broadcast_transaction_with_callback, but won't be held up unless broadcast_transaction_with_callback is called, or the cleanup thread is running.